proposals/0204-slashable-event-verification.md

```diff
@@ -0,0 +1,313 @@
+
+This is the only protocol change that clients need to implement. The remaining
+proposal describes the function of this program, hereafter referred to as the
+slashing program.
+
+### Slashing Program
+
+This slashing program supports two instructions `DuplicateBlockProof`, and
+`CloseProofReport`.
+
+`DuplicateBlockProof` requires 1 account and the `Instructions` sysvar:
+
+0. `proof_account`, expected to be previously initialized with the proof data.
+1. `instructions`, Instructions sysvar
+
+`DuplicateBlockProof` has an instruction data of 48 bytes, containing:
+
+- `0x00`, a fixed-value byte acting as the instruction discriminator
+- `offset`, an unaligned eight-byte little-endian unsigned integer indicating
+  the offset from which to read the proof
+- `slot`, an unaligned eight-byte little-endian unsigned integer indicating the
+  slot in which the violation occured
+- `node_pubkey`, an unaligned 32 byte array representing the public key of the
+  node which committed the violation
+
+We expect the contents of the `proof_account` when read from `offset` to
+deserialize to two byte arrays representing the duplicate shreds.
+The first 4 bytes correspond to the length of the first shred, and the 4 bytes
+after that shred correspond to the length of the second shred.
+
+```rust
+struct DuplicateBlockProofData {
+  shred1_length: u32      // Unaligned four-byte little-endian unsigned integer,
+  shred1: &[u8]           // `shred1_length` bytes representing a shred,
+  shred2_length: u32      // Unaligned four-byte little-endian unsigned integer,
+  shred2: &[u8]           // `shred2_length` bytes representing a shred,
+}
+```
+
+Users are expected to populate the `proof_account` themselves, using an onchain
+program such as the Record program.
+
+`DuplicateBlockProof` aborts if:
+
+- The difference between the current slot and `slot` is greater than 1 epoch's
+  worth of slots as reported by the `Clock` sysvar
+- `offset` is larger than the length of `proof_account`
+- `proof_account[offset..]` does not deserialize cleanly to a
+  `DuplicateBlockProofData`.
+- The resulting shreds do not adhere to the Solana shred format [\[2\]](#notes)
+  or are legacy shred variants.
+- The resulting shreds specify a slot that is different from `slot`.
+- The resulting shreds specify different shred versions.
+
+After deserialization the slashing program will attempt to verify the proof, by
+checking that `shred1` and `shred2` constitute a valid duplicate block proof for
+`slot` and are correctly signed by `node_pubkey`. This is similar to logic used
+in Solana's gossip protocol to verify duplicate block proofs for use in fork choice.
+
+#### Proof verification
+
+`shred1` and `shred2` constitute a valid duplicate block proof if any of the
+following conditions are met:
+
+- Both shreds specify the same index and shred type, however their payloads
+  differ
+- Both shreds specify the same FEC set, however their merkle roots differ
```

Comment:
will there be a feature gate to deprecate / drop all shreds except chained merkles? that would simplify this check to just the comparing signature (`memcmp` 64 bytes vs. 1228) which could add up

Reply:
> That could work since we already validate the merkle root -> signature as part of the sigverify section.

yep exactly, and actually meant that would also obviate the need to compare the payloads with two shreds with the same index as well. so could potentially merge these two conditions into a single "Both shreds specify the same FEC set index, however their signatures differ"

(assuming only chained merkle)

Reply:
the merkle root is only 32 bytes, so I don't think it saves us anything here vs comparing the signature. It requires some compute to put together, but we have to do it anyway for the signature verification

Reply:
discussed offline about the payload check:
- we cannot remove it because a leader could specify 2 shreds with the same index but different fec set indices bypassing the merkle root check
- Once we have fixed size fec sets, this becomes simple to address as we know the range of indices per fec sets
- the payload check is not technically wrong, just inefficient. We can remove it once we have fixed size fec sets